runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -298,15 +298,14 @@ function M.on_diagnostic(_, result, ctx, config)
   local uri = ctx.params.textDocument.uri
   local fname = vim.uri_to_fname(uri)
 
-  if result == nil then
-    return
-  end
-
-  if result.kind == 'unchanged' then
-    return
+  local diagnostics = {}
+  if result ~= nil then
+    if result.kind == 'unchanged' then
+      return
+    end
+    diagnostics = result.items
   end
 
-  local diagnostics = result.items
   if #diagnostics == 0 and vim.fn.bufexists(fname) == 0 then
```

Comment:
Following your logic, if `result == nil`, we'll still have `#diagnostics = 0` so this function would immediately return as it did previously?